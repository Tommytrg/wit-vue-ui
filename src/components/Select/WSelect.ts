export type Option = {key: string, label: string, icon: string | null}
export enum DropdownPosition {
  top = 'top',
  bottom = 'bottom'
}
export const dropdownPostions = Object.values(DropdownPosition)