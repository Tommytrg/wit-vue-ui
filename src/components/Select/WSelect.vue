<template>
  <vSelect
    v-model="model"
    transition="dropdown"
    ref="select"
    :options="options"
    :clearable="false"
    :filterable="false"
    :searchable="false"
    class="language-selector"
  >
    <template #selected-option-container="{ option }">
      <span class="vs__selected">
        <component
          :is="(option as Option).icon"
          v-if="(option as Option).icon"
          class="dropdown-items"
        />
        {{ (option as Option).label }}
      </span>
    </template>

    <template #option="option">
      <div class="option">
        <component
          :is="(option as Option).icon"
          v-if="(option as Option).icon"
          class="dropdown-items"
        />
        <span>{{ (option as Option).label }}</span>
      </div>
    </template>
  </vSelect>
</template>
<script setup lang="ts">
import vSelect from 'vue-select'
import 'vue-select/dist/vue-select.css'
import { dropdownPostions, DropdownPosition} from './WSelect'
import { PropType, computed, DefineComponent, SVGAttributes } from 'vue'

type Option = { key: string; label: string; icon: DefineComponent<SVGAttributes> | null }

const model = defineModel()
const props = defineProps({
  options: {
    type: Array as PropType<Array<Option>>,
    required: true,
  },
  dropdownPosition: {
    type: String as PropType<DropdownPosition>,
    default: DropdownPosition.Bottom,
    validator(value: DropdownPosition) {
      return dropdownPostions.includes(value)
    }
  },
  hexPrimaryColor: {
    type: String,
    default: '#000000'
  },
  hexSecondaryColor: {
    type: String,
    default: '#ffffff'
  }
})

const position = computed(() => props.dropdownPosition === DropdownPosition.Top ? '-210%' : '110%')
</script>

<style lang="scss">

.dropdown-enter {
  transform: translateY(-8px);
}

.dropdown-enter-to {
  transform: translateY(0px);
}
.dropdown-enter-from {
  opacity: 0;
  transform: translateY(-8px);
}

.dropdown-leave {
    transform: translateY(0px);
}

.dropdown-leave-to {
  opacity: 0;
  transform: translateY(-8px);
}

.dropdown-leave-from {
    transform: translateY(0);
}

.dropdown-leave-active {
    transition: all .2s cubic-bezier(.42,0,.58,1);
    transition-delay: .5s;
}

.dropdown-enter-active {
    transition: all .2s cubic-bezier(.42,0,.58,1);
    transition-delay: .5s;
}
.vs--open {
  .vs__open-indicator {
    transform: rotate(180deg) scale(0.5) !important;
  }
}
.language-selector {
  @apply bg-[v-bind(hexPrimaryColor)] w-min px-md py-xs rounded-full flex justify-center;
  .option {
    @apply grid grid-cols-[max-content_max-content] justify-start items-center h-max;
  }
  .vs__dropdown-toggle {
    @apply grid grid-flow-col gap-xs justify-center items-center;
    padding: 0;
    .vs__selected-options {
      padding: 0;
    }
  }
  .vs__dropdown-toggle,
  .vs__dropdown-menu {
    @apply rounded-lg;
    background: v-bind(hexPrimaryColor);
    box-shadow: none;
    border: none;
    color: v-bind(hexSecondaryColor);
    min-width: max-content;
    .dropdown-items {
      margin-right: 8px;
    }
  }
  .vs__dropdown-menu {
    z-index: -10;
    top: v-bind(position);
    left: 0px;
    background: v-bind(hexSecondaryColor);
    border: 2px solid v-bind(hexPrimaryColor);
    border-radius: 16px;
  }
  .vs__open-indicator {
    font-size: 8px;
  }

  .vs__dropdown-option {
    font-family: 'NeueMachina-Regular', sans-serif;
    color: v-bind(hexPrimaryColor)
  }

  .vs__dropdown-option--highlight {
    z-index: 10;
    color: v-bind(hexSecondaryColor);
    background: v-bind(hexPrimaryColor)
  }

  .vs__clear,
  .vs__open-indicator {
    fill: v-bind(hexSecondaryColor);
    transform: scale(0.5);
    transition: transform 150ms cubic-bezier(1, -0.115, 0.975, 0.855);
    transition-timing-function: cubic-bezier(1, -0.115, 0.975, 0.855);
  }

  .vs__selected {
    color: v-bind(hexSecondaryColor);
    margin: 0;
    padding: 0;
    display: flex;
    justify-content: flex-end;
    align-items: center;
  }
  &:hover {
    .vs__selected {
      color: v-bind(hexSecondaryColor);
    }
    .vs__open-indicator {
      color: v-bind(hexSecondaryColor);
    }
  }
  // remove extra space
  .vs__search {
    padding: 0 !important;
  }
  .vs__actions {
    padding: 0;
  }
}

// avoid decrease size on open
.vs--single.vs--open .vs__selected {
  opacity: 1 !important;
  position: inherit !important;
}

// @media (max-width: 706px) {
//   .language-selector {
//     padding: 8px 8px;
//     .vs__selected {
//       width: 100px;
//     }
//     .vs__dropdown-menu {
//       top: -80px;
//       left: 24px;
//     }
//   }
// }
</style>

