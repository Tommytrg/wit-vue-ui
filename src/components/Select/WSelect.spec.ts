import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import WSelect from './WSelect.vue'
import { DropdownPosition } from './WSelect'
import Icon from '@/assets/svg/navigation-cursor.svg?component'

describe('WSelect', () => {
  it('renders properly', () => {
    const wrapper = mount(WSelect, {
      emits: ['update:modelValue'],
      props: {
        dropdownPosition: DropdownPosition.bottom,
        modelValue: {
          key: 'en',
          label: 'English',
          icon: null
        },
        'onUpdate:modelValue': (e) => wrapper.setProps({ modelValue: e }),
        options: [{
            key: 'en',
            label: 'English',
            icon: null
          },{
            key: 'es',
            label: 'Español',
            icon: null,
          }
        ]
      },
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('renders properly an icon when provided', () => {
    const wrapper = mount(WSelect, {
      emits: ['update:modelValue'],
      props: {
        dropdownPosition: DropdownPosition.bottom,
        modelValue: {
          key: 'en',
          label: 'English',
          icon: Icon
        },
        'onUpdate:modelValue': (e) => wrapper.setProps({ modelValue: e }),
        options: [{
            key: 'en',
            label: 'English',
            icon: Icon
          },{
            key: 'es',
            label: 'Español',
            icon: Icon,
          }
        ]
      },
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
