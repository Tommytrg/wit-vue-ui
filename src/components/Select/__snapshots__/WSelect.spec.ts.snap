// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WSelect > renders properly 1`] = `
<div
  class="v-select vs--single vs--unsearchable language-selector"
  dir="auto"
>
  
  
  <div
    aria-expanded="false"
    aria-label="Search for option"
    aria-owns="vs1__listbox"
    class="vs__dropdown-toggle"
    id="vs1__combobox"
    role="combobox"
  >
    <div
      class="vs__selected-options"
    >
      
      
      <span
        class="vs__selected"
      >
        <!--v-if-->
         English
      </span>
      
      
      
      <input
        aria-autocomplete="list"
        aria-controls="vs1__listbox"
        aria-labelledby="vs1__combobox"
        autocomplete="off"
        class="vs__search"
        readonly=""
        type="search"
        value=""
      />
      
    </div>
    <div
      class="vs__actions"
    >
      <button
        aria-label="Clear Selected"
        class="vs__clear"
        style="display: none;"
        title="Clear Selected"
        type="button"
      >
        <svg
          height="10"
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.895455 5l2.842897-2.842898c.348864-.348863.348864-.914488 0-1.263636L9.106534.261648c-.348864-.348864-.914489-.348864-1.263636 0L5 3.104545 2.157102.261648c-.348863-.348864-.914488-.348864-1.263636 0L.261648.893466c-.348864.348864-.348864.914489 0 1.263636L3.104545 5 .261648 7.842898c-.348864.348863-.348864.914488 0 1.263636l.631818.631818c.348864.348864.914773.348864 1.263636 0L5 6.895455l2.842898 2.842897c.348863.348864.914772.348864 1.263636 0l.631818-.631818c.348864-.348864.348864-.914489 0-1.263636L6.895455 5z"
          />
        </svg>
      </button>
      
      <svg
        class="vs__open-indicator"
        height="10"
        role="presentation"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.211364 7.59931l4.48338-4.867229c.407008-.441854.407008-1.158247 0-1.60046l-.73712-.80023c-.407008-.441854-1.066904-.441854-1.474243 0L7 5.198617 2.51662.33139c-.407008-.441853-1.066904-.441853-1.474243 0l-.737121.80023c-.407008.441854-.407008 1.158248 0 1.600461l4.48338 4.867228L7 10l2.211364-2.40069z"
        />
      </svg>
      
      
      <div
        class="vs__spinner"
        style="display: none;"
      >
        Loading...
      </div>
      
    </div>
  </div>
  <transition-stub
    appear="false"
    css="true"
    name="dropdown"
    persisted="false"
  >
    <ul
      id="vs1__listbox"
      role="listbox"
      style="display: none; visibility: hidden;"
    />
  </transition-stub>
  
  
</div>
`;

exports[`WSelect > renders properly an icon when provided 1`] = `
<div
  class="v-select vs--single vs--unsearchable language-selector"
  dir="auto"
>
  
  
  <div
    aria-expanded="false"
    aria-label="Search for option"
    aria-owns="vs2__listbox"
    class="vs__dropdown-toggle"
    id="vs2__combobox"
    role="combobox"
  >
    <div
      class="vs__selected-options"
    >
      
      
      <span
        class="vs__selected"
      >
        <svg
          class="dropdown-items"
          fill="none"
          height="13"
          viewBox="0 0 12 13"
          width="12"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11.17 6.39 5.89 9.44.61 12.48V.29l5.28 3.05z"
            fill="#00E2ED"
          />
        </svg>
         English
      </span>
      
      
      
      <input
        aria-autocomplete="list"
        aria-controls="vs2__listbox"
        aria-labelledby="vs2__combobox"
        autocomplete="off"
        class="vs__search"
        readonly=""
        type="search"
        value=""
      />
      
    </div>
    <div
      class="vs__actions"
    >
      <button
        aria-label="Clear Selected"
        class="vs__clear"
        style="display: none;"
        title="Clear Selected"
        type="button"
      >
        <svg
          height="10"
          width="10"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M6.895455 5l2.842897-2.842898c.348864-.348863.348864-.914488 0-1.263636L9.106534.261648c-.348864-.348864-.914489-.348864-1.263636 0L5 3.104545 2.157102.261648c-.348863-.348864-.914488-.348864-1.263636 0L.261648.893466c-.348864.348864-.348864.914489 0 1.263636L3.104545 5 .261648 7.842898c-.348864.348863-.348864.914488 0 1.263636l.631818.631818c.348864.348864.914773.348864 1.263636 0L5 6.895455l2.842898 2.842897c.348863.348864.914772.348864 1.263636 0l.631818-.631818c.348864-.348864.348864-.914489 0-1.263636L6.895455 5z"
          />
        </svg>
      </button>
      
      <svg
        class="vs__open-indicator"
        height="10"
        role="presentation"
        width="14"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.211364 7.59931l4.48338-4.867229c.407008-.441854.407008-1.158247 0-1.60046l-.73712-.80023c-.407008-.441854-1.066904-.441854-1.474243 0L7 5.198617 2.51662.33139c-.407008-.441853-1.066904-.441853-1.474243 0l-.737121.80023c-.407008.441854-.407008 1.158248 0 1.600461l4.48338 4.867228L7 10l2.211364-2.40069z"
        />
      </svg>
      
      
      <div
        class="vs__spinner"
        style="display: none;"
      >
        Loading...
      </div>
      
    </div>
  </div>
  <transition-stub
    appear="false"
    css="true"
    name="dropdown"
    persisted="false"
  >
    <ul
      id="vs2__listbox"
      role="listbox"
      style="display: none; visibility: hidden;"
    />
  </transition-stub>
  
  
</div>
`;
